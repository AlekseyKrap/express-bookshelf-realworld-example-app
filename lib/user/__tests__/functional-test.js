'use strict';

const chance = require('chance').Chance('user');
const request = require('supertest');
const app = require('../../app');
const User = require('../model');

describe('user', () => {
  let user;

  beforeEach(async () => {
    user = await new User({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: 'password',
      username: chance.word(),
    }).save();
  });

  afterEach(async () => {
    await user.destroy();
  });

  describe('POST /api/users', () => {
    describe('with invalid parameters', () => {
      it('should return a 422', async () => {
        const response = await request(app)
          .post('/api/users')
          .set('Accept', 'application/json')
          .send({user: {}});

        expect(response.statusCode).toBe(422);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('with valid parameters', () => {
      it('should return a 201', async () => {
        const email = chance.email();
        const password = chance.string({length: 8});
        const username = chance.word();

        const response = await request(app)
          .post('/api/users')
          .set('Accept', 'application/json')
          .send({
            user: {
              email,
              password,
              username,
            },
          });

        expect(response.statusCode).toBe(201);

        // It's possible to verify the response via a snapshot here, but it
        // would mean ensuring the same token is generated each time. That would
        // likely involve mocking in order to generate consistent payload data
        // (specifically user id and issued at time), which is something we're
        // deliberately trying to avoid.
        expect(response.body).toEqual({
          user: {
            bio: null,
            email,
            image: null,
            token: expect.any(String),
            username,
          },
        });
        expect(response.body.user.token).toBeAValidJWT({username});
      });
    });
  });

  describe('POST /api/users/login', () => {
    describe('with invalid parameters', () => {
      it('should return a 422', async () => {
        const response = await request(app)
          .post('/api/users/login')
          .set('Accept', 'application/json')
          .send({user: {}});

        expect(response.statusCode).toBe(422);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('with valid parameters', () => {
      it('should return a 200', async () => {
        const response = await request(app)
          .post('/api/users/login')
          .set('Accept', 'application/json')
          .send({
            user: {
              email: user.get('email'),
              password: 'password',
            },
          });

        expect(response.statusCode).toBe(200);
        expect(response.body).toEqual({
          user: {
            bio: user.get('bio'),
            email: user.get('email'),
            image: user.get('image'),
            token: expect.any(String),
            username: user.get('username'),
          },
        });
        expect(response.body.user.token).toBeAValidJWT({
          username: user.get('username'),
        });
      });
    });
  });

  describe('GET /api/user', () => {
    describe('with invalid token', () => {
      it('should return a 401', async () => {
        const response = await request(app)
          .get('/api/user')
          .set('Accept', 'application/json')
          .set('Authorization', 'Token 123')
          .send();

        expect(response.statusCode).toBe(401);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('with valid token', () => {
      it('should return a 200', async () => {
        const response = await request(app)
          .get('/api/user')
          .set('Accept', 'application/json')
          .set('Authorization', `Token ${user.generateJWT()}`)
          .send();

        expect(response.statusCode).toBe(200);
        expect(response.body).toEqual({
          user: {
            bio: user.get('bio'),
            email: user.get('email'),
            image: user.get('image'),
            token: expect.any(String),
            username: user.get('username'),
          },
        });
        expect(response.body.user.token).toBeAValidJWT({
          username: user.get('username'),
        });
      });
    });
  });

  describe('PUT /api/user', () => {
    describe('with invalid token', () => {
      it('should return a 401', async () => {
        const response = await request(app)
          .put('/api/user')
          .set('Accept', 'application/json')
          .set('Authorization', 'Token 123')
          .send();

        expect(response.statusCode).toBe(401);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('with valid parameters', () => {
      it('should return a 200', async () => {
        const response = await request(app)
          .put('/api/user')
          .set('Accept', 'application/json')
          .set('Authorization', `Token ${user.generateJWT()}`)
          .send({
            user: {
              bio: 'Bleep bloop',
              image: 'http://fillmurray.com/200/300',
            },
          });

        expect(response.statusCode).toBe(200);
        expect(response.body).toEqual({
          user: {
            bio: 'Bleep bloop',
            email: user.get('email'),
            image: 'http://fillmurray.com/200/300',
            token: expect.any(String),
            username: user.get('username'),
          },
        });
        expect(response.body.user.token).toBeAValidJWT({
          username: user.get('username'),
        });
      });
    });
  });
});
