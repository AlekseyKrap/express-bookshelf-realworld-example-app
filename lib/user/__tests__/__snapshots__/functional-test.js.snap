// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user GET /api/user with invalid token should return a 401 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "code": "invalid_token",
      "data": Object {
        "UnauthorizedError": Array [
          "jwt malformed",
        ],
      },
      "inner": Object {
        "message": "jwt malformed",
        "name": "JsonWebTokenError",
      },
      "isBoom": true,
      "isServer": false,
      "message": "jwt malformed",
      "name": "UnauthorizedError",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Unauthorized",
          "message": "jwt malformed",
          "statusCode": 401,
        },
        "statusCode": 401,
      },
      "status": 401,
    },
    "message": "jwt malformed",
  },
}
`;

exports[`user POST /api/users with invalid parameters should return a 422 1`] = `
Object {
  "errors": Object {
    "email": Array [
      "can't be blank",
    ],
    "password": Array [
      "can't be blank",
    ],
    "username": Array [
      "can't be blank",
    ],
  },
}
`;

exports[`user POST /api/users/login with invalid parameters should return a 422 1`] = `
Object {
  "errors": Object {
    "email or password": Array [
      "is invalid",
    ],
  },
}
`;

exports[`user PUT /api/user with invalid token should return a 401 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "code": "invalid_token",
      "data": Object {
        "UnauthorizedError": Array [
          "jwt malformed",
        ],
      },
      "inner": Object {
        "message": "jwt malformed",
        "name": "JsonWebTokenError",
      },
      "isBoom": true,
      "isServer": false,
      "message": "jwt malformed",
      "name": "UnauthorizedError",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Unauthorized",
          "message": "jwt malformed",
          "statusCode": 401,
        },
        "statusCode": 401,
      },
      "status": 401,
    },
    "message": "jwt malformed",
  },
}
`;
